### PR TITLE
Use configure_file() to configure version only

### PR DESCRIPTION
**Pull request by trivialfis**
@hcho3  This keeps most of your clean up, but added back the intrin detection code in order to avoid writing `build_config.h` in include directory.  After this, only version upgrade can change that file.

### CODE REVIEW

**Comment by hcho3** [comment]
@trivialfis If the goal is to avoid polluting the include directory, can we apply something like dmlc/dmlc-core#551?

**Comment by hcho3** [comment]
> After this, only version upgrade can change that file

We should formally enforce this, rather than making it an unwritten convention. Perhaps create a separate config step just for the version (`version_config.h`)

**Comment by trivialfis** [comment]
@hcho3 Good point.

**Comment by hcho3** [comment]
Let me try incorporating ideas from dmlc/dmlc-core#551 and dmlc/dmlc-core#566 and get back to you.

**Comment by trivialfis** [comment]
@hcho3 I removed the `build_config.h` completely and moved those macros into `base.h` since that's where most of our macros defined, also previously `base.h` includes `build_config.h`.

I also added a `version_config.h` that does only configuring version information.

File: include/xgboost/base.h [hunk 1]
```diff
@@ -232,7 +246,4 @@ using XGBoostVersionT = int32_t;
 #endif  // DMLC_USE_CXX11 && defined(__GNUC__) && !defined(__clang_version__)
 }  // namespace xgboost
 
-/* Always keep this #include at the bottom of xgboost/base.h */
-#include <xgboost/build_config.h>
```
Comment (hcho3) [inline include/xgboost/base.h]: Can we include `xgboost/version_config.h` here? Otherwise, we'll have to fret about which header gets included first.

File: src/common/hist_util.cc [hunk 2]
```diff
@@ -2,14 +2,16 @@
...
-#include <rabit/rabit.h>
 #include <dmlc/omp.h>
+
+#include <rabit/rabit.h>
 #include <numeric>
 #include <vector>
 
+#include "xgboost/base.h"
 #include "../common/common.h"
+#include "./hist_util.h"
```
Comment (hcho3) [inline src/common/hist_util.cc]: This re-arrangement of headers can be avoided by including `version_config.h` in `base.h`.

File: src/data/data.cc [hunk 3]
```diff
@@ -2,15 +2,16 @@
...
-#include "./sparse_page_writer.h"
-#include "./simple_dmatrix.h"
-#include "./simple_csr_source.h"
+#include "xgboost/data.h"
+#include "xgboost/logging.h"
+#include "xgboost/version_config.h"
+#include "sparse_page_writer.h"
+#include "simple_dmatrix.h"
+#include "simple_csr_source.h"
+
```
Comment (hcho3) [inline src/data/data.cc]: This re-arrangement of headers can be avoided by including `version_config.h` in `base.h`.

File: src/common/hist_util.cc (hunk 2 above)
Comment (trivialfis) [reply src/common/hist_util.cc]: @hcho3 I will remove some re-arrangement of headers.  The `base.h` is included here for mm prefetch , not for version.

File: src/c_api/c_api.cc [hunk 4]
```diff
@@ -16,6 +10,12 @@
 #include <string>
 #include <memory>
 
+#include "xgboost/data.h"
+#include "xgboost/learner.h"
+#include "xgboost/c_api.h"
+#include "xgboost/logging.h"
+#include "xgboost/version_config.h"
```
Comment (hcho3) [inline src/c_api/c_api.cc]: Why is this change necessary?

File: src/data/data.cc (hunk 3 above)
Comment (trivialfis) [reply src/data/data.cc]: @hcho3 It might be better to include what you use.  The linter checks for this convention for standard headers, but not for others.

File: src/c_api/c_api.cc (hunk 4 above)
Comment (trivialfis) [reply src/c_api/c_api.cc]: @hcho3 Just couldn't help myself when adding `version_config.h` here (we added some consistency around `include ""` and `include <>` when exporting c++ header).  I can revert this as unnecessary change.

File: include/xgboost/base.h (hunk 1 above)
Comment (trivialfis) [reply include/xgboost/base.h]: @hcho3 I don't think version have any effect to the rest of the macros right?

**Comment by hcho3** [comment]
Note: Should we need to use `configure_file()` for other things, we should try to apply lessons learned from dmlc/dmlc-core#551 and dmlc/dmlc-core#566.

**Comment by trivialfis** [comment]
@hcho3 Agreed. until we can remove those makefiles, your method works better.

**Comment by hcho3** [comment]
Will merge once tests pass.

**Comment by trivialfis** [comment]
@hcho3 Thanks for the help. ;-)
